components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManager.java

```diff
@@ -284,53 +293,93 @@ public String getUsernameByClaims(Map<String, String> claims, String tenantDomai
+                resultedUserList.addAll(abstractUserStoreManager.getUserListWithID(operationalCondition, domain,
+                        UserCoreConstants.DEFAULT_PROFILE, 2, 1, null, null));
+                if (resultedUserList.size() > 1) {
                     if (log.isDebugEnabled()) {
-                        String message = String.format("No users matched for claim : %s for value : %s", key, value);
-                        log.debug(message);
+                        log.debug("Multiple users matched for given claims set : " +
+                                Arrays.toString(resultedUserList.toArray()));
                     }
-                    return StringUtils.EMPTY;
+                    throw Utils.handleClientException(
+                            IdentityRecoveryConstants.ErrorMessages.ERROR_CODE_MULTIPLE_MATCHING_USERS, null);
                 }
             }
-        }
-        // Return matched user.
-        if (ArrayUtils.isNotEmpty(resultedUserList) && resultedUserList.length == 1) {
-            return resultedUserList[0];
-        } else {
+            // Return empty when no users are found.
+            if (resultedUserList.isEmpty()) {
+                return StringUtils.EMPTY;
+            }
+            // When the code reaches here there only be single user match.
+            return resultedUserList.get(0).getDomainQualifiedUsername();
+        } catch (org.wso2.carbon.user.core.UserStoreException e) {
             if (log.isDebugEnabled()) {
-                log.debug("Multiple users matched for given claims set : " + Arrays.toString(resultedUserList));
+                log.debug("Error while getting users from user store for the given claim set: " +
+                        Arrays.toString(claims.keySet().toArray()));
             }
-            throw Utils
-                    .handleClientException(IdentityRecoveryConstants.ErrorMessages.ERROR_CODE_MULTIPLE_MATCHING_USERS,
-                            null);
+            throw new IdentityRecoveryException(e.getErrorCode(), e.getMessage(), e);
+        } catch (UserStoreException e) {
+            throw new IdentityRecoveryException(e.getMessage(), e);
         }
     }
 
+    /**
+     * Get the expression conditions for the claim set.
+     *
+     * @param claims   List of UserClaims
+     * @param domain   User store domain
+     * @param claimManager   Claim manager
+     * @return expressionConditionList of the claims.
+     * @throws UserStoreException Error while get attribute name.
+     */
+    private List<ExpressionCondition> getExpressionConditionList (Map<String, String> claims, String domain,
+                                                                 ClaimManager claimManager)
+            throws UserStoreException {
+
+        List<ExpressionCondition> expressionConditionList = new ArrayList<>();
+        for (String key : claims.keySet()) {
+            String attributeName = claimManager.getAttributeName(domain, key);
```

Comment:
If we need both key and value to be consumed during the loop, loop on the entry set instead of keys

Reply:
Resolved by https://github.com/wso2-extensions/identity-governance/pull/623/commits/efe78e3032ad8640c0e75a3e32db69809d6ceb05